core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -0,0 +1,1870 @@
+  }
+}
+
+object GlobalPSBTMap extends PSBTMapFactory[GlobalPSBTRecord, GlobalPSBTMap] {
+
+  override def recordFactory: Factory[GlobalPSBTRecord] = GlobalPSBTRecord
+  override def constructMap(elements: Vector[GlobalPSBTRecord]): GlobalPSBTMap =
+    GlobalPSBTMap(elements)
+}
+
+case class InputPSBTMap(elements: Vector[InputPSBTRecord])
+    extends PSBTMap[InputPSBTRecord] {
+  import org.bitcoins.core.psbt.InputPSBTRecord._
+  import org.bitcoins.core.psbt.PSBTInputKeyId._
+
+  def nonWitnessOrUnknownUTXOOpt: Option[NonWitnessOrUnknownUTXO] = {
+    getRecords(NonWitnessUTXOKeyId).headOption
+  }
+
+  def witnessUTXOOpt: Option[WitnessUTXO] = {
+    getRecords(WitnessUTXOKeyId).headOption
+  }
+
+  def partialSignatures: Vector[PartialSignature] = {
+    getRecords(PartialSignatureKeyId)
+  }
+
+  def sigHashTypeOpt: Option[SigHashType] = {
+    getRecords(SigHashTypeKeyId).headOption
+  }
+
+  def redeemScriptOpt: Option[RedeemScript] = {
+    getRecords(RedeemScriptKeyId).headOption
+  }
+
+  def witnessScriptOpt: Option[WitnessScript] = {
+    getRecords(WitnessScriptKeyId).headOption
+  }
+
+  def BIP32DerivationPaths: Vector[BIP32DerivationPath] = {
+    getRecords(BIP32DerivationPathKeyId)
+  }
+
+  def finalizedScriptSigOpt: Option[FinalizedScriptSig] = {
+    getRecords(FinalizedScriptSigKeyId).headOption
+  }
+
+  def finalizedScriptWitnessOpt: Option[FinalizedScriptWitness] = {
+    getRecords(FinalizedScriptWitnessKeyId).headOption
+  }
+
+  def proofOfReservesCommitmentOpt: Option[ProofOfReservesCommitment] = {
+    getRecords(ProofOfReservesCommitmentKeyId).headOption
+  }
+
+  private def getRecords(key: PSBTInputKeyId): Vector[key.RecordType] = {
+    super.getRecords(key, PSBTInputKeyId)
+  }
+
+  def isFinalized: Boolean =
+    getRecords(FinalizedScriptSigKeyId).nonEmpty || getRecords(
+      FinalizedScriptWitnessKeyId).nonEmpty
+
+  /** Finalizes this input if possible, returning None if not */
+  def finalize(input: TransactionInput): Try[InputPSBTMap] = {
+    if (isFinalized) {
+      Success(this)
+    } else {
+      val scriptPubKeyToSatisfyT: Try[ScriptPubKey] = {
+        val witnessUTXOOpt =
+          getRecords(WitnessUTXOKeyId).headOption
+        val nonWitnessUTXOOpt =
+          getRecords(NonWitnessUTXOKeyId).headOption
+
+        (witnessUTXOOpt, nonWitnessUTXOOpt) match {
+          case (None, None) =>
+            Failure(
+              new IllegalStateException(
+                s"Cannot finalize without UTXO record: $this"))
+          case (Some(witnessUtxo), _) =>
+            Success(witnessUtxo.spentWitnessTransaction.scriptPubKey)
+          case (_, Some(utxo)) =>
+            val outputs = utxo.transactionSpent.outputs
+            Try(outputs(input.previousOutput.vout.toInt).scriptPubKey)
+        }
+      }
+
+      scriptPubKeyToSatisfyT.flatMap(finalize)
+    }
+  }
+
+  /** Finalizes this input if possible, returning None if not */
```

Comment:
scaladoc needs update, it's a `Try` not `None`

Reply:
@nkohen